src/Microsoft.AspNet.OData.Shared/Extensions/ODataPathExtensions.cs

```diff
@@ -11,46 +11,75 @@
 
 namespace Microsoft.AspNet.OData.Extensions
 {
-    internal static class ODataPathExtensions
+    /// <summary>
+    /// Extensions method for <see cref="ODataPath"/>.
+    /// </summary>
+    public static class ODataPathExtensions
     {
+        /// <summary>
+        /// Get keys from the last <see cref="KeySegment"/>.
+        /// </summary>
+        /// <param name="path"><see cref="ODataPath"/>.</param>
+        /// <returns>Dictionary of keys.</returns>
         public static Dictionary<string, object> GetKeys(this ODataPath path)
         {
             if (path == null)
             {
                 throw Error.ArgumentNull(nameof(path));
             }
 
-            Dictionary<string, object> keys = new Dictionary<string, object>();
+            List<ODataPathSegment> pathSegments = path.AsList();
+            int position = path.Count - 1;
+            ODataPathSegment pathSegment = pathSegments[position];
 
-            // Books(1)/Authors(1000)/Namespace.SpecialAuthor
-            if (path.LastSegment is TypeSegment)
+            while (!(pathSegment is KeySegment) && position >= 0)
             {
-                ODataPath pathWithoutLastSegmentCastType = path.TrimEndingTypeSegment();
+                pathSegment = pathSegments[--position];
+            }
+
+            return ODataPathHelper.GetKeysFromKeySegment(pathSegment as KeySegment);
+        }
 
-                if (pathWithoutLastSegmentCastType.LastSegment is KeySegment)
-                {
-                    keys = GetKeysFromKeySegment(pathWithoutLastSegmentCastType.LastSegment as KeySegment);
-                }
+        /// <summary>
+        /// Return the last segment in the path, which is not a <see cref="TypeSegment"/> or <see cref="KeySegment"/>.
+        /// </summary>
+        /// <param name="path">The <see cref="ODataPath"/>.</param>
+        /// <returns>An <see cref="ODataPathSegment"/>.</returns>
+        public static ODataPathSegment GetLastNonTypeNonKeySegment(this ODataPath path)
+        {
+            if (path == null)
+            {
+                throw Error.ArgumentNull(nameof(path));
             }
-            // Books(1)/Authors/Namespace.SpecialAuthor/(1000)
-            else if (path.LastSegment is KeySegment)
+
+            if (path.Count == 1)
             {
-                keys = GetKeysFromKeySegment(path.LastSegment as KeySegment);
+                return path.LastSegment;
```

Comment:
How does this statement guarantee that `LastSegment` is neither a `KeySegment` nor a `TypeSegment`? Do you have a test for this scenario?

Reply:
Fixed code and added test

Reply:
I'd still like to understand this line... Is our confidence based on the position that whenever there's only one segment in `path` it cannot be a `TypeSegment` or `KeySegment`?

Reply:
Logically, if we have one segment in the path, it should either be a `SingletonSegment` or `EntitySetSegment`. However we can validate this since ODataPath segments can be constructed manually